### PR TITLE
auto generate uuids for incrememnts() columns

### PR DESCRIPTION
**Pull request by matthewjumpsoffbuildings**
As discussed in this issue https://github.com/colopl/laravel-spanner/issues/133 - initially was thinking using a bit-reversed sequence for auto-incrementing ids would be the optimal solution, but really i think UUIDs are probably better anyway, and more in line with how spanner operates. For example, I believe the bit reversed sequence generator is not as performant as auto generated UUIDs at scale

Added tests, all pass on the latest Spanner emulator

### CODE REVIEW

**Comment by taka-oyama** [comment]
> I believe the bit reversed sequence generator is not as performant as auto generated UUIDs at scale
Do you have anything to backup this claim?

**Comment by matthewjumpsoffbuildings** [comment]
> > I believe the bit reversed sequence generator is not as performant as auto generated UUIDs at scale
> 
> Do you have anything to backup this claim?

I have had email and video call contact with some engineers on the google Spanner team, as I was having some issues at scale with load tests, and they suggested that. Also if you consider how the bit reversed sequences are generated, from [the docs
](https://cloud.google.com/spanner/docs/primary-key-default-value#bit-reversed-sequence)
> A bit-reversed sequence is a schema object that produces a sequence of integers and bit-reverses them. This object uses bit reversal on a private, internal Spanner counter to ensure uniqueness...
> In Spanner, you use SEQUENCE DDL statements along with the bit_reversed_positive attribute to create, alter, or drop a sequence that produces bit-reversed positive values...
> Each sequence maintains a set of internal counters and uses them to generate a value. The sequence counter provides the input to the bit-reversing algorithm.

Unlike UUIDs, which are not state dependent, and are just randomly generated by Spanner, bit-reversed sequences hold a set of counters which are incremented by Spanner internally, which can be less performant when experiencing a high volume of writes

**Comment by taka-oyama** [comment]
Did you try the load test with `GENERATE_UUID`?

**Comment by matthewjumpsoffbuildings** [comment]
Hmm the unit test is failing here, but on my local Spanner emulator it was running fine? I am guessing the Dockerfile or something needs updating to the latest Spanner?

**Comment by matthewjumpsoffbuildings** [comment]
> Did you try the load test with `GENERATE_UUID`?

Yes and it was more performant. The most performant option is to generate the UUID client side, but still if one is going to use the auto-generated columns, I think UUID is a better choice

**Comment by taka-oyama** [comment]
Dockerfile is fine. It pulled the latest emulator.

Please check the error. There is something wrong with your code.

```

1) Colopl\Spanner\Tests\Schema\BlueprintTest::test_increments
ErrorException: Undefined array key "large"

/project/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:254
/project/tests/Schema/BlueprintTest.php:599

ERRORS!
Tests: 137, Assertions: 456, Errors: 1, Skipped: 2.
2
```

**Comment by matthewjumpsoffbuildings** [comment]
Yes, my mistake, sorry, pushed a new commit

**Comment by taka-oyama** [comment]
> if one is going to use the auto-generated columns, I think UUID is a better choice

That's a fair point. Especially if bit reversed sequence generator has a scaling issue.

File: src/Schema/Blueprint.php [hunk 1]
```diff
@@ -18,8 +18,10 @@
 namespace Colopl\Spanner\Schema;
 
 use Colopl\Spanner\Concerns\MarksAsNotSupported;
+use Illuminate\Database\Query\Expression;
 use Illuminate\Database\Schema\Blueprint as BaseBlueprint;
 use Illuminate\Database\Schema\ColumnDefinition;
+use Illuminate\Support\Facades\DB;
```
Comment (taka-oyama) [inline src/Schema/Blueprint.php:24]: This doesn't look like it's used?

```suggestion
```

**Comment by matthewjumpsoffbuildings** [comment]
> > if one is going to use the auto-generated columns, I think UUID is a better choice
> 
> That's a fair point. Especially if bit reversed sequence generator has a scaling issue.

I would say bit rev seq is still performant, just not as optimal as UUID

**Comment by taka-oyama** [comment]
Closing since I have created a similar PR [here](https://github.com/colopl/laravel-spanner/pull/175) which solves the same problem.

**Pull request closed without merging**